massa-ledger/src/ledger_db.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2022 MASSA LABS <info@massa.net>
+
+use massa_hash::Hash;
+use massa_models::Address;
+use rocksdb::{ColumnFamilyDescriptor, Options, WriteBatch, DB};
+
+use crate::{ledger_changes::LedgerEntryUpdate, LedgerEntry, SetOrDelete, SetOrKeep};
+
+const DB_PATH: &str = "../.db";
+const LEDGER_CF: &str = "ledger";
+const METADATA_CF: &str = "metadata";
+const OPEN_ERROR: &str = "critical: rocksdb open operation failed";
+const CRUD_ERROR: &str = "critical: rocksdb crud operation failed";
+const CF_ERROR: &str = "critical: rocksdb column family operation failed";
+
+/// Ledger sub entry enum
+pub enum LedgerSubEntry {
+    /// Balance
+    Balance,
+    /// Bytecode
+    Bytecode,
+    /// Datastore entry
+    Datastore(Hash),
+}
+
+pub(crate) struct LedgerDB(DB);
+
+// NOTE: still handle separate bytecode for now to avoid too many refactoring at once
+macro_rules! bytecode_key {
+    ($addr:ident) => {
+        format!("b:{}", $addr).as_bytes()
+    };
+}
+
+macro_rules! data_key {
+    ($addr:ident, $key:ident) => {
+        format!("{}:{}", $addr, $key).as_bytes()
+    };
+}
+
+impl LedgerDB {
+    pub fn new() -> Self {
+        // db options
+        let mut db_opts = Options::default();
+        db_opts.create_if_missing(true);
+        db_opts.create_missing_column_families(true);
+
+        // database init
+        let db = DB::open_cf_descriptors(
+            &db_opts,
+            DB_PATH,
+            vec![
+                ColumnFamilyDescriptor::new(LEDGER_CF, Options::default()),
+                ColumnFamilyDescriptor::new(METADATA_CF, Options::default()),
+            ],
+        )
+        .expect(OPEN_ERROR);
+
+        // return database
+        LedgerDB(db)
+    }
+
+    pub fn put(&mut self, addr: &Address, ledger_entry: LedgerEntry) {
+        let mut batch = WriteBatch::default();
+        let key = addr.to_bytes();
+        let handle = self.0.cf_handle(LEDGER_CF).expect(CF_ERROR);
+
+        // balance
+        batch.put_cf(
+            handle,
+            key,
+            ledger_entry.parallel_balance.to_raw().to_be_bytes(),
+        );
+
+        // bytecode
+        batch.put_cf(handle, bytecode_key!(addr), ledger_entry.bytecode);
+
+        // datastore
+        for (hash, entry) in ledger_entry.datastore {
+            batch.put_cf(handle, data_key!(addr, hash), entry);
+        }
+
+        // write batch
+        self.0.write(batch).expect(CRUD_ERROR);
+    }
+
+    pub fn update(&mut self, addr: &Address, entry_update: LedgerEntryUpdate) {
+        let mut batch = WriteBatch::default();
+        let key = addr.to_bytes();
+        let handle = self.0.cf_handle(LEDGER_CF).expect(CF_ERROR);
+
+        // balance
+        if let SetOrKeep::Set(balance) = entry_update.parallel_balance {
+            batch.put_cf(handle, key, balance.to_raw().to_be_bytes());
```

Comment:
Can we save varint instead ? To have a more light ledger.

Reply:
Yeah I'm pretty sure we can.